libs/core/gasnet_base/src/gasnet_environment.cpp

```diff
@@ -0,0 +1,655 @@
+} done_t;
+
+static void AM_signal(gasnet_token_t token, gasnet_handlerarg_t a0, gasnet_handlerarg_t a1) {
+  done_t* done = (done_t*) get_ptr_from_args(a0, a1);
+  uint_least32_t prev;
+  prev = done->count.fetch_add(1, std::memory_order_seq_cst);
+  if (prev + 1 == done->target)
+    done->flag = 1;
+}
+
+static void AM_signal_long(gasnet_token_t token, void *buf, size_t nbytes,
+                           gasnet_handlerarg_t a0, gasnet_handlerarg_t a1) {
+  done_t* done = (done_t*) get_ptr_from_args(a0, a1);
+  uint_least32_t prev;
+  prev = done->count.fetch_add(1, std::memory_order_seq_cst);
+  if (prev + 1 == done->target)
+    done->flag = 1;
+}
+
+// Put from arg->src (which is local to the AM handler) back to
+// arg->dst (which is local to the caller of this AM).
+// nbytes is < gasnet_AMMaxLongReply here (see chpl_comm_get).
+//
+static void AM_reply_put(gasnet_token_t token, void* buf, size_t nbytes) {
+   xfer_info_t* x = static_cast<xfer_info_t*>(buf);
+
+   assert(nbytes == sizeof(xfer_info_t));
+
+   GASNET_Safe(gasnet_AMReplyLong2(token, SIGNAL_LONG,
+      x->src, x->size, x->tgt,
+      Arg0(x->ack), Arg1(x->ack)));
+}
+
+// Copy from the payload in this active message to dst.
+//
+static
+void AM_copy_payload(gasnet_token_t token, void* buf, size_t nbytes,
+   gasnet_handlerarg_t ack0, gasnet_handlerarg_t ack1,
+   gasnet_handlerarg_t dst0, gasnet_handlerarg_t dst1)
+{
+   void* dst = get_ptr_from_args(dst0, dst1);
+   {
+      // would prefer to protect the memory segments
+      // associated with each node (n-node mutex)
+      // will require future work
+      //
+      std::lock_guard<hpx::mutex> lk(hpx::util::gasnet_environment::dshm_mut);
+      std::memcpy(dst, buf, nbytes);
+   }
+
+   GASNET_Safe(gasnet_AMReplyShort2(token, SIGNAL, ack0, ack1));
+}
+
+static gasnet_handlerentry_t ftable[] = {
+   {SIGNAL,        (void (*)())&AM_signal},
+   {SIGNAL_LONG,   (void (*)())&AM_signal_long},
+   {DO_REPLY_PUT,  (void (*)())&AM_reply_put},
+   {DO_COPY_PAYLOAD, (void (*)())&AM_copy_payload}
+};
+
+//
+// Initialize one of the above.
+//
+static inline
+void init_done_obj(done_t* done, int target) {
+  done->count.store(0, std::memory_order_seq_cst);
+  done->target = target;
+  done->flag = 0;
+}
+
+static inline void am_poll_try() {
+  // Serialize polling for IBV, UCX, Aries, and OFI. Concurrent polling causes
+  // contention in these configurations. For other configurations that are
+  // AM-based (udp/amudp, mpi/ammpi) serializing can hurt performance.
+  //
+#if defined(GASNET_CONDUIT_IBV) || defined(GASNET_CONDUIT_UCX) || defined(GASNET_CONDUIT_ARIES) || defined(GASNET_CONDUIT_OFI)
+  std::lock_guard<hpx::spinlock> lk(gasnet_environment::pollingLock);
+    (void) gasnet_AMPoll();
+#else
+    (void) gasnet_AMPoll();
+#endif
+}
+
+static inline
+void wait_done_obj(done_t* done, bool do_yield)
+{
+  while (!done->flag) {
+    am_poll_try();
+    if (do_yield)
+      hpx::this_thread::yield();
```

Comment:
needs to be changed to `hpx::execution_base::this_thread::yield();`

Reply:
I don't think so. the code looks fine as it is

Reply:
I hit a compilation issue with the udp target (gcc-11). The compiler couldn't resolve yield as a member of this_thread. Not sure what the story is there.

Reply:
You might just miss the header file declaring this function (`#include <hpx/thread.hpp>`).

Reply:
@hkaiser yep, that was the problem, wrong header file.